### ISSUE TITLE
Some info that may help ...

### ISSUE THREAD

**Issue by dunky11** [lucidrains/natural-speech-pytorch#1]
Thank you for trying to implement this. I tried to implement natural speech myself but I got stuck at the soft-dtw part. I wrote with the authors and they told me some stuff that may help and isn't obvious from reading the paper.

1. For calculating the soft-dtw losses the most straightforward approach is to calculate a matrix of shape (batch_size, n_channels, n_phones, n_spectrogram_frames). Obviously, this matrix is huge and will blow up the GPU, so they told me that they "use a for loop to calculate each channel and add them together." This probably involved some low-level stuff.
2. This may be obvious for you but it wasn't for me, in the Durator the softmax "is calculated over the phoneme dimension".
3. Since during inference the length n corresponding to the number of spectrogram frames in the upsampling layer is unknown, it "is calculated from the predicted duration from duration predictor".
4. Discriminators are exactly the same as Multi-Period-Discriminators from HiFiGAN, so they can be taken from the [official repo](https://github.com/jik876/hifi-gan/blob/master/models.py).
5. The Normalizing Flow, Posterior Encoder, and Decoder seem to be the same as VITS, [here is the official repo](https://github.com/jaywalnut310/vits).

This may help, good luck with the implementation :)

**Comment by lucidrains** [comment]
@dunky11 Hi Tim, and thank you for all the advice

I'm new to the audio domain, so this project will probably take me a while to figure out

I was also wondering what your thoughts on https://arxiv.org/abs/2206.08297 were

**Comment by lucidrains** [comment]
@dunky11 without taking a close look at soft-dtw, are you aware of pykeops? https://github.com/getkeops/keops

**Comment by dunky11** [comment]
@lucidrains I wasn't aware of that library, I looked into the [torch.Function](https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function) but then one has to write the backward pass himself. Anything that takes a loop and does some parallelization while also automatically calculating the gradients would be great. 
Regarding the paper, I haven't seen that before, I'm not super into audio language modelling.

**Comment by lucidrains** [comment]
@dunky11 ohh got it Tim, no problem :smile: will start on this soon!
